src/s3/listobjects.cpp

```diff
@@ -56,90 +60,206 @@ asio::awaitable<void> irods::s3::actions::handle_listobjects_v2(
-        //        std::cout<<"Adding it to the document"<<std::endl;
-        document.add_child("ListBucketResult.Contents", object);
-    }
-    // Required for genquery limitations :p
-    query = fmt::format(
-        "select COLL_NAME,DATA_NAME,DATA_OWNER_NAME,DATA_SIZE where COLL_NAME like '{}/{}'",
-        resolved_path.parent_path().c_str(),
-        filename_prefix);
-    std::cout << query << std::endl;
-    for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
-        found_objects = true;
-        ptree object;
-        object.put("Key", i[0].substr(base_length) + "/" + i[1]);
-        object.put("Etag", i[0] + "/" + i[1]);
-        object.put("Owner", i[2]);
-        object.put("Size", atoi(i[3].c_str()));
-        document.add_child("ListBucketResult.Contents", object);
-    }
 
-    if (found_objects) {
-        std::stringstream s;
-        boost::property_tree::xml_parser::xml_writer_settings<std::string> settings;
-        settings.indent_char = ' ';
-        settings.indent_count = 4;
-        std::cout << "Found objects" << std::endl;
-        boost::property_tree::write_xml(s, document, settings);
-        beast::http::response<beast::http::string_body> response;
-        response.body() = s.str();
-        std::cout << s.str();
+    if (delimiter_in_request) {
 
-        beast::http::write(socket, response);
-    }
-    else {
-        std::cout << "Couldn't find anything" << std::endl;
-        beast::http::response<beast::http::empty_body> response;
-        response.result(beast::http::status::not_found);
-        beast::http::write(socket, response);
+        if (full_path.object_name().string() == "") {
+            // Path ends in a slash, this is an exact collection match
+            // and all objects in that collection
+            
+            // Get exact collections underneath this collection 
+            query = fmt::format(
+                "select COLL_NAME where COLL_NAME like '{}/%' and COLL_NAME not like '{}/%/%'",
+                full_path.parent_path().string(),
+                full_path.parent_path().string());
+            std::cout << query << std::endl;
+            for (auto&& i : irods::query<RcComm>(thing.get(), query)) {
+                found_objects = true;
+                ptree object;
+                std::string key = (i[0].size() > base_length ? i[0].substr(base_length) : "");
+                if (key.rfind("/", 0) == 0) {
+                    key = key.substr(1);
+                }
+                key += "/";
+                object.put("Prefix", key);
+                document.add_child("ListBucketResult.CommonPrefixes", object);
+            }
+
+            // Get the data objects within the collection
+            query = fmt::format(
+                "select COLL_NAME, DATA_NAME, DATA_OWNER_NAME, DATA_SIZE, DATA_MODIFY_TIME where COLL_NAME = '{}'",
+                full_path.parent_path().string());
```

Comment:
`.string()` -> `.c_str()`

Reply:
Same as earlier.  This is a path object not a std::string.

Reply:
`full_path.parent_path().c_str()`

Reply:
done